mainnet-rehearsals/3-R4-upgrade-2-of-2/README.md

```diff
@@ -0,0 +1,174 @@
+# Rehearsal 4 - Protocol Upgrade via Nested Multisig
+
+## Objective
+
+In this rehearsal we will be performing a protocol upgrade of the
+Superchain.
+
+Once Completed:
+1. The L1ERC721BridgeProxy contract will be upgraded to a new
+   implementation.
+2. The OptimismPortalProxy contract will be upgraded to a new
+   implementation and reinitialized.
+
+The call that will be executed by the Safe contract is defined in a
+json file. This will be the standard approach for all transactions.
+
+Note that no onchain actions will be taking place during this
+signing. You won’t be submitting a transaction and your address
+doesn’t even need to be funded. These are offchain signatures produced
+with your wallet which will be collected by a Facilitator, who will
+submit all signatures and perform the execution onchain.
+
+
+## Approving the transaction
+
+### 1. Move to the appropriate folder for this rehearsal task:
+
+```
+cd superchain-ops/mainnet-rehearsals/4-upgrade-2-of-2
+```
+
+### 2. Setup Ledger
+
+Your Ledger needs to be connected and unlocked. The Ethereum
+application needs to be opened on Ledger with the message "Application
+is ready".
+
+### 3. Simulate and validate the transaction
+
+Make sure your ledger is still unlocked and run the following.
+
+``` shell
+just simulate-council
+```
+
+You will see a "Simulation link" from the output.
+
+Paste this URL in your browser. A prompt may ask you to choose a
+project, any project will do. You can create one if necessary.
+
+Click "Simulate Transaction".
+
+We will be performing 3 validations and extract the domain hash and
+message hash to approve on your Ledger:
+
+1. Validate integrity of the simulation.
+2. Validate correctness of the state diff.
+3. Validate and extract domain hash and message hash to approve.
+
+#### 3.1. Validate integrity of the simulation.
+
+Make sure you are on the "Overview" tab of the tenderly simulation, to
+validate integrity of the simulation, we need to
+
+1. "Network": Check the network is Ethereum Mainnet.
+2. "Timestamp": Check the simulation is performed on a block with a
+   recent timestamp (i.e. close to when you run the script).
+3. "Sender": Check the address shown is your signer account. If not,
+   you will need to determine which “number” it is in the list of
+   addresses on your ledger. By default the script will assume the
+   derivation path is m/44'/60'/0'/0/0. By calling the script with
+   `just simulate-council 1` it will derive the address using
+   m/44'/60'/1'/0/0 instead.
+
+![](./images/tenderly-overview-network.png)
+
+#### 3.2. Validate correctness of the state diff.
+
+Now click on the "State" tab. Verify that
```

Comment:
I found this section somewhat difficult to use while actually performing the validation. I took a shot at a refactoring that is (at least for me) easier to follow. 

The key changes are: 
1. the top level list item starts with the address to inspect, which is more similar to the tenderly layout. Accordingly, the `OptimismPortalProxy` changes are now all in a single list item.
2. applying code formatting on addresses and variable names 

My updated verification list: 

```
#### 3.2. Validate correctness of the state diff.

Now click on the "State" tab. Verify that:

1. In `0xce479c1299adbe12f34a479f404937151af19d92` (the `L1ERC721BridgeProxy` contract):
   1. The implementation of the `L1ERC721BridgeProxy` contract (storage key
    [0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc](https://github.com/ethereum-optimism/optimism/blob/cb42a6108d780451f6cecceff8182e11aa6a0490/packages/contracts-bedrock/src/libraries/Constants.sol#L27C9-L27C75))
    (`0xce479c1299adbe12f34a479f404937151af19d92`) is changed to
    `0x3311aC7F72bb4108d9f4D5d50E7623B1498A9eC0`.
2. In `0xDd3840F548Bd8fA41421050ed7C28753B494734c` (the `OptimismPortalProxy` contract):
   1. The implementation (storage key is the same as above) is changed to
   `0x0A9d47e531825FaaA2863D4d10DC8E5E0B91BfB0` ([source
   code](https://github.com/ethereum-optimism/optimism/blob/716f81a6fc4ef125364b95a799474082ea3eb062/packages/contracts-bedrock/src/L1/OptimismPortal.sol#L123)).
   1. storage slot `0x0` (`_initialized`) is set to `0x2`.
   2. storage slot `0x1` (`ResourceParams`) is set to
       `0x000000000117814800000000000000000000000000000000000000003b9aca00`,
       which means `prevBlockNum` is set to 0x1178148, prevBoughtGas is
       set to 0, and `prevBaseFee` is set to 0x3b9aca00.
   3. storage slot `0x32` (`l2Sender`) set to `0xdead`.
   4. storage slot `0x35` (`l2Oracle`) set to `0x0000000000000000000000dfe97868233d1aa22e815a266982f2cf17685a2700`.
   5. storage slot `0x36` (`systemConfig`) set to `0x000000000000000000000000229047fed2591dbec1ef1118d64f7af3db9eb290`.
   6. storage slot `0x37` (`guardian`) set to `0x0000000000000000000000009ba6e03d8b90de867373db8cf1a58d2f7f006b3a`.
```